File: components/esp_board_manager/peripherals/periph_adc/periph_adc.c
/*
 * SPDX-FileCopyrightText: 2025 Espressif Systems (Shanghai) CO., LTD
 * SPDX-License-Identifier: LicenseRef-Espressif-Modified-MIT
 *
 * See LICENSE file for details.
 */

#include "esp_log.h"
#include <string.h>
#include "esp_board_manager_defs.h"
#include "esp_board_periph.h"
#include "periph_adc.h"

static const char *TAG = "PERIPH_ADC";

static inline bool adc_unit_supported(adc_unit_t unit)
{
#ifdef SOC_ADC_DIG_SUPPORTED_UNIT
    return SOC_ADC_DIG_SUPPORTED_UNIT(unit);
#else
    return unit == ADC_UNIT_1;
#endif  /* SOC_ADC_DIG_SUPPORTED_UNIT */
}

static inline esp_err_t continuous_adc_init(periph_adc_config_t *adc_cfg, periph_adc_handle_t *handle)
{
    esp_err_t err = ESP_FAIL;
    adc_continuous_handle_t continuous_handle = NULL;

    err = adc_continuous_new_handle(&adc_cfg->cfg.continuous.handle_cfg, &continuous_handle);
    if (err != ESP_OK) {
        ESP_LOGE(TAG, "Create adc continuous handle failed: %s", esp_err_to_name(err));
        return err;
    }

    adc_continuous_config_t dig_cfg = {
        .sample_freq_hz = adc_cfg->cfg.continuous.sample_freq_hz,
        .conv_mode = adc_cfg->cfg.continuous.conv_mode,
        .format = adc_cfg->cfg.continuous.format,
    };
    int pattern_num = adc_cfg->cfg.continuous.pattern_num;
    if (pattern_num <= 0 || pattern_num > SOC_ADC_PATT_LEN_MAX) {
        ESP_LOGE(TAG, "Invalid pattern_num:%d", pattern_num);
        adc_continuous_deinit(continuous_handle);
        return ESP_ERR_INVALID_ARG;
    }
    adc_digi_pattern_config_t adc_pattern[SOC_ADC_PATT_LEN_MAX] = {0};
    dig_cfg.pattern_num = pattern_num;
    if (adc_cfg->cfg.continuous.cfg_mode == PERIPH_ADC_CONTINUOUS_CFG_MODE_SINGLE_UNIT) {
        for (int i = 0; i < pattern_num; i++) {
            adc_pattern[i].atten = adc_cfg->cfg.continuous.cfg.single_unit.atten;
            adc_pattern[i].channel = adc_cfg->cfg.continuous.cfg.single_unit.channel_id[i];
            adc_pattern[i].unit = adc_cfg->cfg.continuous.cfg.single_unit.unit_id;
            adc_pattern[i].bit_width = adc_cfg->cfg.continuous.cfg.single_unit.bit_width;
        }
    } else if (adc_cfg->cfg.continuous.cfg_mode == PERIPH_ADC_CONTINUOUS_CFG_MODE_PATTERN) {
        memcpy(adc_pattern, adc_cfg->cfg.continuous.cfg.patterns,
               pattern_num * sizeof(adc_digi_pattern_config_t));
    } else {
        ESP_LOGE(TAG, "Invalid cfg_mode:%d", adc_cfg->cfg.continuous.cfg_mode);
        adc_continuous_deinit(continuous_handle);
        return ESP_ERR_INVALID_ARG;
    }
    dig_cfg.adc_pattern = adc_pattern;
    for (int i = 0; i < pattern_num; i++) {
        if (!adc_unit_supported((adc_unit_t)adc_pattern[i].unit)) {
            ESP_LOGE(TAG, "ADC unit %u not supported in continuous mode", adc_pattern[i].unit);
            adc_continuous_deinit(continuous_handle);
            return ESP_ERR_NOT_SUPPORTED;
        }
        if (adc_pattern[i].channel >= SOC_ADC_CHANNEL_NUM((adc_unit_t)adc_pattern[i].unit)) {
            ESP_LOGE(TAG, "Invalid ADC channel %u for unit %u", adc_pattern[i].channel, adc_pattern[i].unit);
            adc_continuous_deinit(continuous_handle);
            return ESP_ERR_INVALID_ARG;
        }
    }

    err = adc_continuous_config(continuous_handle, &dig_cfg);
    if (err != ESP_OK) {
        ESP_LOGE(TAG, "Configure continuous adc handle failed: %s", esp_err_to_name(err));
        adc_continuous_deinit(continuous_handle);
        return err;
    }
    handle->continuous = continuous_handle;
    ESP_LOGI(TAG, "Create adc continuous handle success");
    return ESP_OK;
}

static inline esp_err_t oneshot_adc_init(periph_adc_config_t *adc_cfg, periph_adc_handle_t *handle)
{
    esp_err_t err = ESP_FAIL;
    adc_oneshot_unit_handle_t oneshot_handle = NULL;

    err = adc_oneshot_new_unit(&adc_cfg->cfg.oneshot.unit_cfg, &oneshot_handle);
    if (err != ESP_OK) {
        ESP_LOGE(TAG, "Create adc oneshot unit failed: %s", esp_err_to_name(err));
        return err;
    }

    adc_channel_t channel = (adc_channel_t)adc_cfg->cfg.oneshot.channel_id;
    err = adc_oneshot_config_channel(oneshot_handle, channel, &adc_cfg->cfg.oneshot.chan_cfg);
    if (err != ESP_OK) {
        ESP_LOGE(TAG, "Configure adc oneshot channel failed: %s", esp_err_to_name(err));
        adc_oneshot_del_unit(oneshot_handle);
        return err;
    }
    handle->oneshot = oneshot_handle;
    ESP_LOGI(TAG, "Create adc oneshot unit success");
    return ESP_OK;
}

int periph_adc_init(void *cfg, int cfg_size, void **periph_handle)
{
    if (!cfg || !periph_handle || cfg_size != sizeof(periph_adc_config_t)) {
        ESP_LOGE(TAG, "Invalid parameters");
        return -1;
    }

    periph_adc_config_t *adc_cfg = (periph_adc_config_t *)cfg;

    periph_adc_handle_t *handle = (periph_adc_handle_t *)calloc(1, sizeof(periph_adc_handle_t));
    if (handle == NULL) {
        ESP_LOGE(TAG, "Failed to allocate memory for periph_adc_handle_t");
        return -1;
    }

    esp_err_t err = ESP_FAIL;
    if (adc_cfg->role == ESP_BOARD_PERIPH_ROLE_CONTINUOUS) {
        err = continuous_adc_init(adc_cfg, handle);
        if (err != ESP_OK) {
            goto cleanup;
        }
    } else if (adc_cfg->role == ESP_BOARD_PERIPH_ROLE_ONESHOT) {
        err = oneshot_adc_init(adc_cfg, handle);
        if (err != ESP_OK) {
            goto cleanup;
        }
    } else {
        ESP_LOGE(TAG, "Unknown role %d", adc_cfg->role);
        goto cleanup;
    }

    *periph_handle = (void *)handle;
    return 0;

cleanup:
    free(handle);
    return -1;
}

int periph_adc_deinit(void *periph_handle)
{
    if (!periph_handle) {
        ESP_LOGE(TAG, "Invalid handle");
        return -1;
    }
    periph_adc_handle_t *handle = (periph_adc_handle_t *)periph_handle;
    const char *periph_name = NULL;
    esp_err_t err = esp_board_periph_get_name_by_handle(periph_handle, &periph_name);
    if (err != ESP_OK) {
        ESP_LOGE(TAG, "Failed to get periph name: %s", esp_err_to_name(err));
        goto cleanup;
    }

    periph_adc_config_t *cfg = NULL;
    err = esp_board_periph_get_config(periph_name, (void **)&cfg);
    if (err != ESP_OK) {
        ESP_LOGE(TAG, "Failed to get config: %s", esp_err_to_name(err));
        goto cleanup;
    }

    if (cfg->role == ESP_BOARD_PERIPH_ROLE_CONTINUOUS) {
        adc_continuous_handle_t continuous_handle = handle->continuous;
        err = adc_continuous_deinit(continuous_handle);
        if (err != ESP_OK) {
            ESP_LOGE(TAG, "Failed to deinit continuous_handle failed: %s", esp_err_to_name(err));
            goto cleanup;
        }
    } else if (cfg->role == ESP_BOARD_PERIPH_ROLE_ONESHOT) {
        adc_oneshot_unit_handle_t oneshot_handle = handle->oneshot;
        err = adc_oneshot_del_unit(oneshot_handle);
        if (err != ESP_OK) {
            ESP_LOGE(TAG, "Failed to del oneshot_handle: %s", esp_err_to_name(err));
            goto cleanup;
        }
    } else {
        ESP_LOGW(TAG, "Unknown role on deinit: %d", cfg->role);
    }

    ESP_LOGI(TAG, "Deinitialize periph_adc role=%d", cfg->role);

    free(handle);
    return 0;

cleanup:
    free(handle);
    return -1;
}

File: components/esp_board_manager/peripherals/periph_adc/periph_adc.py
# SPDX-FileCopyrightText: 2025 Espressif Systems (Shanghai) CO., LTD
# SPDX-License-Identifier: LicenseRef-Espressif-Modified-MIT
#
# See LICENSE file for details.

# ADC device config parser
VERSION = 'v1.0.0'

import logging

from generators.adc_channel_mapper import adc_channels_to_gpios, adc_patterns_to_gpios

logger = logging.getLogger(__name__)

# Define valid ADC configuration values
VALID_ADC_ROLES = ['continuous', 'oneshot']
# Map string roles to enum values
ADC_ROLE_MAP = {
    'continuous': 'ESP_BOARD_PERIPH_ROLE_CONTINUOUS',
    'oneshot': 'ESP_BOARD_PERIPH_ROLE_ONESHOT'
}
VALID_ADC_UNITS = ['ADC_UNIT_1', 'ADC_UNIT_2']
VALID_ADC_ATTEN = [
    'ADC_ATTEN_DB_0',
    'ADC_ATTEN_DB_2_5',
    'ADC_ATTEN_DB_6',
    'ADC_ATTEN_DB_12',
    'ADC_ATTEN_DB_11'
]
VALID_ADC_BITWIDTH = [
    'ADC_BITWIDTH_9',
    'ADC_BITWIDTH_10',
    'ADC_BITWIDTH_11',
    'ADC_BITWIDTH_12',
    'ADC_BITWIDTH_13',
    'ADC_BITWIDTH_DEFAULT',
    'SOC_ADC_DIGI_MAX_BITWIDTH'
]
VALID_ADC_FORMATS = [
    'ADC_DIGI_OUTPUT_FORMAT_TYPE1',
    'ADC_DIGI_OUTPUT_FORMAT_TYPE2'
]
VALID_ADC_CONV_MODES = [
    'ADC_CONV_SINGLE_UNIT_1',
    'ADC_CONV_SINGLE_UNIT_2',
    'ADC_CONV_BOTH_UNIT',
    'ADC_CONV_ALTER_UNIT'
]
VALID_ADC_ULP_MODES = [
    'ADC_ULP_MODE_DISABLE',
    'ADC_ULP_MODE_FSM',
    'ADC_ULP_MODE_RISCV',
    'ADC_ULP_MODE_LP_CORE'
]

MAX_ADC_PATTERN_ENTRIES = 16

def validate_enum_value(value: str, enum_type: str, valid_values: list) -> bool:
    """Validate if the enum value is within the valid range.

    Args:
        value: The enum value to validate
        enum_type: The type of enum for error messages
        valid_values: List of valid values

    Returns:
        True if valid, False otherwise
    """
    if value not in valid_values:
        print(f"❌ YAML VALIDATION ERROR: Invalid {enum_type} value '{value}'!")
        print(f'   File: board_peripherals.yaml')
        print(f'   📋 Peripheral: ADC configuration')
        print(f"   ❌ Invalid value: '{value}'")
        print(f'   ✅ Valid values: {valid_values}')
        print(f'   ℹ️ Please use one of the valid enum values listed above')
        return False
    return True

def get_enum_value(value: str, default_value: str, enum_type: str, valid_values: list) -> str:
    """Helper function to get enum value with validation.

    Args:
        value: The value from YAML config
        default_value: The default value to use if value is not provided
        enum_type: The type of enum for validation
        valid_values: List of valid values

    Returns:
        The enum value to use
    """
    if not value:
        result = default_value
    else:
        result = value

    # Validate the enum value
    if not validate_enum_value(result, enum_type, valid_values):
        raise ValueError(f"Invalid {enum_type} value '{value}'. Please use one of the valid enum values.")

    return result

def get_adc_bit_width(value, default_value):
    """Get ADC bit width from YAML while preserving enum strings or numeric widths."""
    result = default_value if value is None else value

    if isinstance(result, str):
        if validate_enum_value(result, 'ADC bit width', VALID_ADC_BITWIDTH):
            return result
    elif isinstance(result, int) and not isinstance(result, bool) and result > 0:
        return result
    else:
        validate_enum_value(result, 'ADC bit width', VALID_ADC_BITWIDTH)

    raise ValueError(
        f"Invalid ADC bit width value '{value}'. "
        'Please use one of the valid enum values or a positive integer bit width.'
    )

def get_includes() -> list:
    """Return list of required include headers for ADC"""
    return [
        'periph_adc.h',
    ]

def parse_adc_continuous_config(continuous_config: dict) -> dict:
    """Parse ADC continuous mode specific configuration"""
    channel_list = continuous_config.get('channel_list')
    legacy_channel_id = continuous_config.get('channel_id')
    if channel_list is not None and legacy_channel_id is not None:
        raise ValueError("ADC continuous config cannot contain both 'channel_list' and legacy 'channel_id'")

    channel_values = channel_list if channel_list is not None else legacy_channel_id
    patterns = continuous_config.get('patterns')
    has_patterns = isinstance(patterns, list) and len(patterns) > 0
    has_channel_list = channel_values is not None
    if has_patterns and has_channel_list:
        raise ValueError("ADC continuous config cannot contain both 'patterns' and 'channel_list'")
    if not has_patterns and not has_channel_list:
        raise ValueError("ADC continuous config requires either 'patterns' or 'channel_list'")

    parsed_patterns = []
    channel_id_array = []
    unit = None
    if has_patterns:
        for item in patterns:
            if not isinstance(item, dict):
                raise ValueError("Each 'patterns' item must be a mapping")
            parsed_patterns.append({
                'unit': get_enum_value(item.get('unit'), 'ADC_UNIT_1', 'ADC unit', VALID_ADC_UNITS),
                'channel': int(item.get('channel', -1)),
                'atten': get_enum_value(item.get('atten'), 'ADC_ATTEN_DB_0', 'ADC attenuation', VALID_ADC_ATTEN),
                'bit_width': get_adc_bit_width(item.get('bit_width'), 'ADC_BITWIDTH_DEFAULT'),
            })
    else:
        if isinstance(channel_values, list) and len(channel_values) <= 0:
            raise ValueError('Empty channel_list for ADC continuous device')
        if not isinstance(channel_values, list):
            channel_values = [channel_values]
        channel_id_array = [int(ch) for ch in channel_values]
        for ch in channel_id_array:
            if ch < 0:
                raise ValueError('channel_list cannot contain negative channels')

    pattern_num = len(parsed_patterns) if has_patterns else len(channel_id_array)
    if pattern_num <= 0:
        raise ValueError('pattern_num must be greater than 0')
    if pattern_num > MAX_ADC_PATTERN_ENTRIES:
        raise ValueError(f'pattern_num too large ({pattern_num}), max {MAX_ADC_PATTERN_ENTRIES}')
    if 'pattern_num' in continuous_config:
        expect = int(continuous_config.get('pattern_num', pattern_num))
        if expect != pattern_num:
            raise ValueError(f'pattern_num({expect}) mismatch with configured entries({pattern_num})')

    explicit_conv_mode = continuous_config.get('conv_mode')
    conv_mode = explicit_conv_mode
    if explicit_conv_mode is not None:
        conv_mode = get_enum_value(
            conv_mode,
            'ADC_CONV_SINGLE_UNIT_1',
            'ADC conversion mode',
            VALID_ADC_CONV_MODES
        )
    elif has_patterns:
        units = {item['unit'] for item in parsed_patterns}
        if units == {'ADC_UNIT_1'}:
            conv_mode = 'ADC_CONV_SINGLE_UNIT_1'
        elif units == {'ADC_UNIT_2'}:
            conv_mode = 'ADC_CONV_SINGLE_UNIT_2'
        else:
            conv_mode = 'ADC_CONV_BOTH_UNIT'
    else:
        unit = get_enum_value(
            continuous_config.get('unit_id'),
            'ADC_UNIT_1',
            'ADC unit',
            VALID_ADC_UNITS
        )
        conv_mode = 'ADC_CONV_SINGLE_UNIT_1' if unit == 'ADC_UNIT_1' else 'ADC_CONV_SINGLE_UNIT_2'

    if explicit_conv_mode is not None:
        if has_patterns:
            units = {item['unit'] for item in parsed_patterns}
            if units == {'ADC_UNIT_1'} and conv_mode != 'ADC_CONV_SINGLE_UNIT_1':
                raise ValueError('conv_mode must be ADC_CONV_SINGLE_UNIT_1 when all patterns use ADC_UNIT_1')
            if units == {'ADC_UNIT_2'} and conv_mode != 'ADC_CONV_SINGLE_UNIT_2':
                raise ValueError('conv_mode must be ADC_CONV_SINGLE_UNIT_2 when all patterns use ADC_UNIT_2')
            if len(units) > 1 and conv_mode in {'ADC_CONV_SINGLE_UNIT_1', 'ADC_CONV_SINGLE_UNIT_2'}:
                raise ValueError('conv_mode cannot be single-unit when patterns span multiple ADC units')
        else:
            if unit is None:
                unit = get_enum_value(
                    continuous_config.get('unit_id'),
                    'ADC_UNIT_1',
                    'ADC unit',
                    VALID_ADC_UNITS
                )
            expected_single = 'ADC_CONV_SINGLE_UNIT_1' if unit == 'ADC_UNIT_1' else 'ADC_CONV_SINGLE_UNIT_2'
            if conv_mode != expected_single:
                raise ValueError(
                    f"conv_mode '{conv_mode}' is inconsistent with unit_id '{unit}'. Expected {expected_single}"
                )

    handle_cfg = {
        'max_store_buf_size': continuous_config.get('max_store_buf_size', 1024),
        'conv_frame_size': continuous_config.get('conv_frame_size', 256),
        'flags.flush_pool': continuous_config.get('flush_pool', 1),
    }

    continuous_cfg = {
        'handle_cfg': handle_cfg,
        'sample_freq_hz': continuous_config.get('sample_freq_hz', 20000),
        'format': get_enum_value(
            continuous_config.get('format'),
            'ADC_DIGI_OUTPUT_FORMAT_TYPE2',
            'ADC format',
            VALID_ADC_FORMATS
        ),
        'conv_mode': conv_mode,
        'pattern_num': pattern_num,
        'cfg_mode': 'PERIPH_ADC_CONTINUOUS_CFG_MODE_PATTERN' if has_patterns else 'PERIPH_ADC_CONTINUOUS_CFG_MODE_SINGLE_UNIT',
    }

    if has_patterns:
        for i, item in enumerate(parsed_patterns):
            if item['channel'] < 0:
                raise ValueError(f'patterns[{i}].channel cannot be negative')
        continuous_cfg['cfg'] = {
            'patterns': parsed_patterns
        }
    else:
        continuous_cfg['cfg'] = {
            'single_unit': {
                'unit_id': get_enum_value(
                    continuous_config.get('unit_id'),
                    'ADC_UNIT_1',
                    'ADC unit',
                    VALID_ADC_UNITS
                ),
                'atten': get_enum_value(
                    continuous_config.get('atten'),
                    'ADC_ATTEN_DB_0',
                    'ADC attenuation',
                    VALID_ADC_ATTEN
                ),
                'bit_width': get_adc_bit_width(continuous_config.get('bit_width'), 'SOC_ADC_DIGI_MAX_BITWIDTH'),
                'channel_id': channel_id_array,
            }
        }

    return continuous_cfg

def parse_adc_oneshot_config(oneshot_config: dict) -> dict:
    """Parse ADC oneshot mode specific configuration"""
    unit_cfg = {
        'unit_id': get_enum_value(
            oneshot_config.get('unit_id'),
            'ADC_UNIT_1',
            'ADC unit',
            VALID_ADC_UNITS
        ),
        'clk_src': oneshot_config.get('clk_src', 'ADC_RTC_CLK_SRC_DEFAULT'),
        'ulp_mode': get_enum_value(
            oneshot_config.get('ulp_mode'),
            'ADC_ULP_MODE_DISABLE',
            'ADC ULP mode',
            VALID_ADC_ULP_MODES
        )
    }

    channel_id = oneshot_config.get('channel_id')
    if channel_id is None:
        raise ValueError(f'Missing channel ID for ADC oneshot device')
    if isinstance(channel_id, list):
        raise ValueError('channel_id for ADC oneshot device must be a single integer, not a list')
    if isinstance(channel_id, bool) or not isinstance(channel_id, int):
        raise ValueError(f"channel_id for ADC oneshot device must be an integer, got '{type(channel_id).__name__}'")
    if channel_id < 0:
        raise ValueError('channel_id for ADC oneshot device cannot be negative')
    chan_cfg = {
        'atten': get_enum_value(
            oneshot_config.get('atten'),
            'ADC_ATTEN_DB_0',
            'ADC attenuation',
            VALID_ADC_ATTEN
        ),
        'bitwidth': get_adc_bit_width(oneshot_config.get('bit_width'), 'SOC_ADC_DIGI_MAX_BITWIDTH')
    }

    oneshot_cfg = {
        'channel_id': channel_id,
        'unit_cfg': unit_cfg,
        'chan_cfg': chan_cfg
    }

    return oneshot_cfg

def parse(name: str, full_config: dict, peripherals_dict=None) -> dict:
    """Parse ADC peripheral configuration from YAML to C structure

    Args:
        name: Peripheral name
        config: Configuration dictionary from YAML

    Returns:
        Dictionary containing ADC configuration structure
    """
    try:
        # Get the actual config from the full_config
        config = full_config.get('config', {})

        # Validate role
        role = full_config.get('role')
        if role is None:
            raise ValueError("Missing ADC role for ADC device '{}'".format(name))
        elif not validate_enum_value(role, 'ADC role', VALID_ADC_ROLES):
            raise ValueError(f"Invalid ADC role '{role}' for ADC device '{name}'")

        # Convert string role to enum value
        enum_role = ADC_ROLE_MAP.get(role)
        if enum_role is None:
            raise ValueError(f"Invalid ADC role '{role}' for ADC device '{name}'")

        # Parse sub-type specific configuration
        if role == 'continuous':
            continuous_cfg = parse_adc_continuous_config(config)
            cfg_union = {'continuous': continuous_cfg}
        else:  # oneshot
            oneshot_cfg = parse_adc_oneshot_config(config)
            cfg_union = {'oneshot': oneshot_cfg}

        # Build the main configuration structure
        struct_init = {
            'role': enum_role,
            'cfg': cfg_union
        }

        # Build the result
        result = {
            'struct_type': 'periph_adc_config_t',
            'struct_var': f'{name}_cfg',
            'struct_init': struct_init,
            '_role': role  # Store role for code generation
        }

        return result

    except ValueError as e:
        # Re-raise ValueError with more context
        raise ValueError(f"YAML validation error in ADC peripheral '{name}': {e}")
    except Exception as e:
        # Catch any other exceptions and provide context
        raise ValueError(f"Error parsing ADC peripheral '{name}': {e}")


def extract_metadata(name: str, raw_config: dict, parse_result: dict, context: dict) -> dict:
    chip_name = context.get('chip')
    struct_init = parse_result.get('struct_init', {})
    cfg_union = struct_init.get('cfg', {})

    if 'oneshot' in cfg_union:
        oneshot_cfg = cfg_union.get('oneshot', {})
        unit_id = oneshot_cfg.get('unit_cfg', {}).get('unit_id', 'ADC_UNIT_1')
        channel_id = oneshot_cfg.get('channel_id')
        if channel_id is None:
            return {'io': {}}
        try:
            mapped_channel = adc_channels_to_gpios(chip_name, unit_id, [channel_id])[0]
        except FileNotFoundError as e:
            logger.warning(
                "Skipping ADC metadata extraction for peripheral '%s' on chip '%s': %s",
                name,
                chip_name,
                e,
            )
            return {'io': {}}
        return {'io': {'channel_id': mapped_channel}}

    continuous_cfg = cfg_union.get('continuous', {})
    cfg_mode = continuous_cfg.get('cfg_mode')
    cfg = continuous_cfg.get('cfg', {})

    if cfg_mode == 'PERIPH_ADC_CONTINUOUS_CFG_MODE_PATTERN':
        try:
            mapped_channels = adc_patterns_to_gpios(chip_name, cfg.get('patterns', []))
        except FileNotFoundError as e:
            logger.warning(
                "Skipping ADC metadata extraction for peripheral '%s' on chip '%s': %s",
                name,
                chip_name,
                e,
            )
            return {'io': {}}
        return {
            'io': {
                'channel': mapped_channels,
            }
        }

    single_unit_cfg = cfg.get('single_unit', {})
    channels = single_unit_cfg.get('channel_id', [])
    if not channels:
        return {'io': {}}

    try:
        mapped_channels = adc_channels_to_gpios(
            chip_name,
            single_unit_cfg.get('unit_id', 'ADC_UNIT_1'),
            channels,
        )
    except FileNotFoundError as e:
        logger.warning(
            "Skipping ADC metadata extraction for peripheral '%s' on chip '%s': %s",
            name,
            chip_name,
            e,
        )
        return {'io': {}}

    return {
        'io': {
            'channel_id': mapped_channels,
        }
    }
